### ISSUE TITLE
Invalid option -:

### ISSUE THREAD

**Issue by agaryen** [rbenv/rbenv#1219]
Hey,

For a little while now, I've got these errors showing up for any command that uses ruby 2.6.5:

`$ ruby`
`Traceback (most recent call last):
ruby: invalid option -:  (-h will show valid options) (RuntimeError)
`

`$ rails s`
`Traceback (most recent call last):
/home/rizzen/.rbenv/versions/2.6.5/bin/ruby: invalid option -:  (-h will show valid options) (RuntimeError)`

I guess this is related to rbenv's forwarding of the command line to the right ruby binary, but I couldn't pinpoint the source of the issue.
It started happening after I installed and used ruby 2.7.0 for another project. My main project is still in 2.6.5 and I can't launch any ruby/rails command from there anymore.

I managed to get rid of this error by removing my ~/.rbenv dir and by rebooting (strangely, even shell reset wasn't effective without a reboot).

Any ideas?

**Comment by mislav** [comment]
Strange! Do you have a way of reproducing this still? I never saw this, and it smells of something specific on your system.

First thing I would check is what gets executed when I run ruby in my shell: `type ruby`; for me it reports that the executable `~/.rbenv/shims/ruby` will get run, which is appropriate for an rbenv setup. Then I can inspect that shim; it ends in:
```sh
exec "$HOME/.rbenv/libexec/rbenv" exec "$program" "$@"
```

You can see that it forwards all the arguments to `rbenv exec ruby`, which forwards all the arguments to the `ruby` executable for the proper version: https://github.com/rbenv/rbenv/blob/e429cc0bd823e08b54bd66636565c8e47ac43828/libexec/rbenv-exec#L47

I don't see what could cause the `ruby: invalid option -` error, but you may be able to debug it by setting `RBENV_DEBUG=1 ruby`.

**Comment by agaryen** [comment]
Hey @mislav,
Thanks for your time.

I managed to get rid of this error one more time, after several hours of bruteforce.
To be honest, I'm not sure it has something to do with rbenv. I suspect it bootstrapped the situation initially, but I later tried to downgrade my system ruby binary to 2.6.5 and got the same error executing it, same with an rvm installed ruby.

I printed $program and $args from the ruby shim, and no sign of any dash or anything abnormal.
`which ruby` always pointed to the ruby shim, and `rbenv which ruby` to the 2.6.5 version.

For some strange reason, only the -h worked. Every other params got me the same error message. 2.7.0 always worked like a charm, meanwhile.

I removed my ~/.rbenv folder, uninstalled every ruby binary and lib on my system, uninstalled rbenv, all of that several times. At some point, 2.6.5 started working again, I can't make sense of it, even looking at my command history..

Since it appears to be this kind of mystical issue, I'd rather not touch anything for the rest of the week and let it work as it does. But I'll probably run into the same issue next time I install ruby 2.7.0 through rbenv like I did last two times. I'll take the time to debug it more thoroughly this time, and will keep you posted.

We can close this issue in the meantime.

**Comment by mislav** [comment]
Thank you for following up!

**Comment by t447** [comment]
I just feel the need to put this here:
https://stackoverflow.com/questions/59851061/only-one-version-of-ruby-works-when-multiple-versions-are-installed

I ran into the same issue and this is what was the problem in my case

> The perpetrator seems to be the echo "export RUBYOPT='-W:no-deprecated -W:no-experimental'" line in your .zsh file, I'm not sure if Ruby 2.6.5 understands it.

And this is what helped me solve it.

>Remove echo "export RUBYOPT='-W:no-deprecated -W:no-experimental'" from your .zsh / .bashrc / .profile and restart your shell.

I had set the variable to remove noise form `rails 6.0.2.1` on `ruby 2.7.0`, the same options got picked up by `ruby 2.6.5` when I switched projects. To get around it I erase the variable (`set -e RUBYOPT` in fish shell) first before running anything in an older ruby version, this still lets me keep the clean logs in other projects running on `2.7.0` .

**Comment by agaryen** [comment]
Yep. I pinpointed the issue to this export too a few days ago. I never thought it could be correlated.

**Comment by mislav** [comment]
Nice find!

**Comment by L0stLink** [comment]
No problem! it is always a good practice to leave solutions on relevant issues, if you come across them. And thank you for all your work.

**Comment by t447** [comment]
No problem! it is always a good practice to leave solutions on relevant issues, if you come across them. And thank you for all your work.

**Comment by wacko** [comment]
Is there a way to set up `RUBYOPT='-W:no-deprecated'` for ruby 2.7+ only?

**Comment by JoshCheek** [comment]
For me, the issue was that I had incorrectly set my ruby with `chruby ruby 2.7.2`, 
I fixed it by changing it to `chruby ruby-2.7.2`

-----

This caused the following **INCORRECT** environment variables:

```sh
$ env | grep RUBYOPT
RUBYOPT=2.7.2

$ bundle exec env | grep RUBYOPT
RUBYOPT=-r/Users/josh/.rubies/ruby-2.7.2/lib/ruby/2.7.0/bundler/setup 2.7.2
BUNDLER_ORIG_RUBYOPT=2.7.2
```

-----

Once I updated that to use the dash, it became **CORRECT** (I assume, given that it worked after this):

```sh
$ env | grep RUBYOPT

$ bundle exec env | grep RUBYOPT
BUNDLER_ORIG_RUBYOPT=BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL
RUBYOPT=-r/Users/josh/.rubies/ruby-2.7.2/lib/ruby/2.7.0/bundler/setup
```

-----

My versions:

```sh
$ bundle -v
Bundler version 2.1.4

$ ruby -v
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [arm64-darwin20]

$ uname -a
Darwin MacBook-Air 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64
```
